mesos-client/src/main/scala/com/mesosphere/mesos/client/Session.scala

```diff
@@ -0,0 +1,184 @@
+import akka.http.scaladsl.model._
+import akka.http.scaladsl.settings.ConnectionPoolSettings
+import akka.stream.Materializer
+import akka.stream.scaladsl.Flow
+import akka.util.Timeout
+import com.typesafe.scalalogging.StrictLogging
+
+import scala.util.{Failure, Success}
+
+/**
+  * A session captures the connection information with Mesos. This is among other things the URL and credentials providers.
+  *
+  * It behaves similar to Python Request's [[https://2.python-requests.org/en/master/user/advanced/#session-objects Session object]].
+  * Thus it provides methods to construct and connect to Mesos.
+  *
+  * @param url The Mesos master URL.
+  * @param streamId The Mesos stream ID. See the [[http://mesos.apache.org/documentation/latest/scheduler-http-api/#calls docs]] for details.
+  * @param authorization A [[CredentialsProvider]] if the connection is secured.
+  */
+case class Session(url: URL, streamId: String, authorization: Option[CredentialsProvider] = None)(
+    implicit askTimout: Timeout) {
+
+  /**
+    * Construct a new [[HttpRequest]] for a serialized Mesos call and a set of authorization, ie session token.
+    * @param bytes The bytes of the serialized Mesos call.
+    * @param maybeCredentials The session token if required.
+    * @return The [[HttpRequest]] with proper headers and body.
+    */
+  def createPostRequest(bytes: Array[Byte], maybeCredentials: Option[HttpCredentials]): HttpRequest = {
+    println(s"Body: ${bytes.map(_.toChar).mkString}")
+    HttpRequest(
+      HttpMethods.POST,
+      uri = Uri(s"$url/api/v1/scheduler"),
+      entity = HttpEntity(MesosClient.ProtobufMediaType, bytes),
+      headers = MesosClient.MesosStreamIdHeader(streamId) :: maybeCredentials.map(Authorization(_)).toList
+    )
+  }
+
+  /** @return A flow that makes Mesos calls and outputs HTTP responses. */
+  def post(implicit system: ActorSystem, mat: Materializer): Flow[Array[Byte], HttpResponse, NotUsed] =
+    authorization match {
+      case Some(credentialsProvider) =>
+        val sessionActor = system.actorOf(SessionActor.props(credentialsProvider, createPostRequest))
+        Flow[Array[Byte]].ask[HttpResponse](1)(sessionActor)
+      case None =>
+        Flow[Array[Byte]].map(createPostRequest(_, None)).via(connection)
+    }
+
+  /**
+    * A connection flow factory that will create a flow that only processes one request at a time.
+    *
+    * It will reconnect if the server closes the connection after a successful request.
+    *
+    * @return A connection flow for single requests.
+    */
+  private def connection(implicit system: ActorSystem, mat: Materializer): Flow[HttpRequest, HttpResponse, NotUsed] = {
+    // Constructs the connection pool settings with defaults and overrides the max connections and pipelining limit so
+    // that only one request at a time is processed. See https://doc.akka.io/docs/akka-http/current/configuration.html
+    // for details.
```

Comment:
```suggestion
    // for details.
    // *IMPORTANT*: DO NOT CHANGE maxConnections OR pipeliningLimit! Otherwise, USI won't guarantee request order to Mesos!
```